File: 2017_a_pratice/week6/H - Patchouli Spell Cards/main.py
C = [[0 for i in range(105)] for i in range(105)]
for i in range(0, 102):
    C[i][0]=1
    for j in range(1, i+1):
        C[i][j] = C[i-1][j] + C[i-1][j-1]

def gcd(a, b):
    if b == 0:
        return a
    return gcd(b, a % b)


while True:
    try:
        m, n, l = map(int, raw_input().split())
        if l > m:
            print('mukyu~')
        else:
            dp = [[0 for i in range(105)] for i in range(105)]
            dp[0][0] = 1
            for i in range(1, n+1):
                for j in range(1, m+1):
                    for k in range(0, min(j+1, l)):
                        dp[i][j] += dp[i-1][j-k]*C[m-(j-k)][k]
            tot = n ** m
            use = 0
            for i in range(1, n+1):
                use += dp[i][m]
            use = tot - use
            g = gcd(tot, use)
            print str(use/g)+'/'+str(tot/g)
    except:
        break

File: 2017_a_pratice/week6/H - Patchouli Spell Cards/Main.java
package a2017.practice.week5;

import java.math.BigInteger;
import java.util.Scanner;

/**
 * Created by chen on 2017/4/13 0013.
 */
public class Main {
    static final int maxn = 105;
    static BigInteger C[][] = new BigInteger[maxn][maxn];
    static BigInteger dp[][] = new BigInteger[maxn][maxn];
    static BigInteger N[] = new BigInteger[maxn];

    private static void Fac(BigInteger X[]) {
        X[0] = BigInteger.ONE;
        for (int i = 1; i < X.length; i++) {
            X[i] = X[i - 1].multiply(BigInteger.valueOf(i));
        }
    }

    private static void Comb(BigInteger X[][]) {
        for (int i = 0; i < maxn; i++) {
            for (int j = 0; j <= i; j++) {
                X[i][j] = N[i].divide(N[j]).divide(N[i - j]);
            }
        }
    }

    public static void main(String args[]) {
        Fac(N);
        Comb(C);
        int n, m, l;
        Scanner cin = new Scanner(System.in);
        while (cin.hasNext()) {
            m = cin.nextInt();
            n = cin.nextInt();
            l = cin.nextInt();
            if (l > m) {
                System.out.println("mukyu~");
                continue;
            }
            for (int i = 0; i < maxn; i++) {
                for (int j = 0; j < maxn; j++) {
                    dp[i][j] = BigInteger.ZERO;
                }
            }
            dp[0][0] = BigInteger.ONE;
            for (int i = 1; i <= n; i++) {
                for (int j = 1; j <= m; j++) {
                    for (int k = 0; k <= Math.min(j, l - 1); k++) {
                        dp[i][j] = dp[i][j].add(dp[i - 1][j - k].multiply(C[m - (j - k)][k]));
                    }
                }
            }
            BigInteger all = BigInteger.valueOf(n).pow(m);
            BigInteger use = BigInteger.ZERO;
            for (int i = 1; i <= n; i++) {
                use = use.add(dp[i][m]);
            }
            use = all.subtract(use);
            BigInteger g = all.gcd(use);
            System.out.println(use.divide(g) + "/" + all.divide(g));
        }
    }
}
